### ISSUE TITLE
[Feature] Client Side Commands

### ISSUE THREAD

**Issue by AterAnimAvis** [MinecraftForge/MinecraftForge#7397]
Opening this issue to get Ideas / Feedback / Suggestions.

Apologies in advance for any broken English (Im fairly bad at writing text).

Im willing to write a PR for this, however I am unsure of what I should actually aim for.

I've got a mostly working prototype (implemented as a mod) at [AterAnimAvis/ClientCommands](https://github.com/AterAnimAvis/ClientCommands), note it is still very much WIP with only a few of the Argument Types ported (And some that need a rework).

### Single Event
A Single Event in `ClientPlayNetHandler#handleCommandList(SCommandListPacket)` to notify Mods when the Vanilla dispatcher gets updated.
	
##### Pros:
- Quick + Simple on Forge Side
- Gets rid of [this hack](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/impl/ClientCommandsImpl.java#L94-L99)
	
##### Cons:
- No [Single Standard](https://xkcd.com/927/) for actually implementing the Commands

### (Full / Partial) System
A Single Event [RegisterClientCommandsEvent](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/event/ClientCommandEvent.java) like [RegisterCommandsEvent](https://github.com/MinecraftForge/MinecraftForge/blob/1.16.x/src/main/java/net/minecraftforge/event/RegisterCommandsEvent.java) to register the Commands.

A Single Event [ClientCommandEvent](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/event/ClientCommandEvent.java) like [CommandEvent](https://github.com/MinecraftForge/MinecraftForge/blob/1.16.x/src/main/java/net/minecraftforge/event/CommandEvent.java) to intercept Client commands before they are executed.

A Counterpart to `Commands` called [ClientCommands](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/ClientCommands.java) with necessary helper methods.

A `ClientCommandSource` counterpart to `CommandSource`.

Note theres a couple of options with the `ClientCommandSource`
- extend `CommandSource` directly
  - Most simple Arguments (`BlockStateArgument`, `ColorArgument`, ...) types will work with no further code on Forge side.
  - Exceptions at runtime with Arguments (`BlockPosArgument`, `TeamArgument`, ...) actually expecting a Server.
- extend `ISuggestionProvider`
  - No Exceptions
  - Need to declare (or loosen) [getters](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/arguments/selector/ClientEnchantmentArgument.java#L14-L16) as most have a parameter type of `CommandContext<CommandSource>`
  - Need to declare [custom implementations](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/arguments/selector/ClientGameProfileArgument.java) for Arguments that expect a Server

The Partial part comes into the fact that we don't have to fully port all argument types at once (Or ever in the case of a few like `FunctionArgument` which is as far as I can tell Server Side only)

##### Pros:
- Unified System

##### Cons:
- Technical Debt

##### Current Impl
Note: Most of the Implementation needs to be cleaned up (It was originally used as part of a unpublished development tool for a [client side only mod](https://github.com/AterAnimAvis/BlockRenderer) where I only required mostly Brigadier ArgumentTypes).

The Mod currently piggybacks on Vanilla's `CommandSuggestionHelper` via modifing the `ClientPlayNetHandler#commandDispatcher`.

There's a seperate Execution Dispatcher that is used to allow [fast bailing](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/impl/ClientCommandsImpl.java#L129-L130) on a non Client-Side Command (i.e. if we don't have `/command` in the execution dispatcher we want to pass the message onto the server) and [slower-bailing](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/impl/ClientCommandsImpl.java#L142-L164) at the cost of emitting `RegisterClientCommandsEvent` twice.

I listen for a `SCommandListPacket` as this is the point where Vanilla updates its suggestions. (Note: one cause of this packet can be a reload of DataPackRegistries)

**Comment by GotoLink** [comment]
See #6952 for my PR with this subject in mind.
It only adds what is absolutely necessary.
Features like ClientCommandEvent and ClientCommandSource have few uses anyway.

**Comment by AterAnimAvis** [comment]
Apologies, Some people on discord mentioned there was need of a pr and I somehow managed to miss yours when I checked the open prs.

Your Implementation is a lot cleaner than my current public POC and probably close to the route I would of chosen (I have 3 partial Implementations in local exploring what the best implementation would of been) when I got feedback on what to aim for.

The only real changes I would do to your PR would be to add the [ClientCommands](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/ClientCommands.java#L20-L34) Utility methods to better match with the Minecraft Command API and change the event name to ClientCommandRegistryEvent.

One thing that might be worth PRing is adding Generics to most of the Argument getters where possible.

Also it’s possible to add to chat history [like so](https://github.com/AterAnimAvis/ClientCommands/blob/master/src/main/java/clientcommands/client/commands/impl/ClientCommandsImpl.java#L257-L259).
